mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/MemoryEventDeadLetters.java

```diff
@@ -48,45 +48,47 @@
         Preconditions.checkArgument(registeredGroup != null, REGISTERED_GROUP_CANNOT_BE_NULL);
         Preconditions.checkArgument(failDeliveredEvent != null, FAIL_DELIVERED_EVENT_CANNOT_BE_NULL);
 
-        deadLetters.put(registeredGroup, failDeliveredEvent);
-        return Mono.empty();
+        return Mono.fromRunnable(() -> deadLetters.put(registeredGroup, failDeliveredEvent))
+            .subscribeWith(MonoProcessor.create())
+            .then();
     }
 
     @Override
     public Mono<Void> remove(Group registeredGroup, Event.EventId failDeliveredEventId) {
         Preconditions.checkArgument(registeredGroup != null, REGISTERED_GROUP_CANNOT_BE_NULL);
         Preconditions.checkArgument(failDeliveredEventId != null, FAIL_DELIVERED_ID_EVENT_CANNOT_BE_NULL);
 
-        deadLetters.get(registeredGroup)
-            .removeIf(event -> event.getEventId().equals(failDeliveredEventId));
-
-        return Mono.empty();
+        return Flux.fromIterable(deadLetters.get(registeredGroup))
+            .filter(event -> event.getEventId().equals(failDeliveredEventId))
+            .next()
+            .doOnNext(event -> deadLetters.remove(registeredGroup, event))
+            .subscribeWith(MonoProcessor.create())
+            .then();
     }
 
     @Override
     public Mono<Event> failedEvent(Group registeredGroup, Event.EventId failDeliveredEventId) {
         Preconditions.checkArgument(registeredGroup != null, REGISTERED_GROUP_CANNOT_BE_NULL);
         Preconditions.checkArgument(failDeliveredEventId != null, FAIL_DELIVERED_ID_EVENT_CANNOT_BE_NULL);
 
-        Optional<Event> failedEventMaybe = deadLetters.get(registeredGroup)
-            .stream()
+        return Flux.fromIterable(deadLetters.get(registeredGroup))
             .filter(event -> event.getEventId().equals(failDeliveredEventId))
-            .findFirst();
-
-        return Mono.justOrEmpty(failedEventMaybe);
+            .next()
+            .subscribeWith(MonoProcessor.create());
     }
 
     @Override
     public Flux<Event.EventId> failedEventIds(Group registeredGroup) {
         Preconditions.checkArgument(registeredGroup != null, REGISTERED_GROUP_CANNOT_BE_NULL);
 
-        return Flux.fromStream(deadLetters.get(registeredGroup)
-            .stream()
-            .map(Event::getEventId));
+        return Flux.fromIterable(deadLetters.get(registeredGroup))
+            .map(Event::getEventId)
+            .subscribeWith(EmitterProcessor.create());
     }
 
     @Override
     public Flux<Group> groupsWithFailedEvents() {
-        return Flux.fromIterable(deadLetters.keySet());
+        return Flux.fromIterable(deadLetters.keySet())
+            .subscribeWith(EmitterProcessor.create());
```

Comment:
What does this do?

Reply:
yes

Reply:
Then we don't need it everywhere we only perform reads.

Reply:
> Then we don't need it everywhere we only perform reads.

did you mean, for reading operations => eagerly subscribe, other modifing operations => let callers subscribe ?

Reply:
When I call a create/delete/update API it makes sens to eagerly subscribe it yes.

For read operations however we can be more "lazy" IMO. If the user don't block / do not read all, then we can save the overhead of reading extra entries. Not doing that eager subscribe would make this choice explicit IMO.

Reply:
okay to me, @blackheaven how do you think?

Reply:
ok for me